tr1/stats: add extra stats from TR2

walkawayy opened the pull request:
Added stats are ammo hits/used, health packs used, and distance travelled.

#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TRX/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change
- [X] I have added a readme entry about my new feature or OG bug fix, or it is a different change

#### Description
Backports stats from TR2 including ammo hits / used, health packs used, and distance travelled. Uses the new unreleased `VERSION_7` savegame header. Also helps prep for some libtrx merges in the future by bringing more feature parity. Hooks into the existing detailed stats section since another stats option seems like overkill.

Some questions / concerns:
1. Attached some screenshots below to show how it looks at different text scalings. Required a slight modification to `ROW_HEIGHT_BARE` in `stats_dialog.c` from `30` to `25` to fit the end of level stats screen at max text scaling. I didn't know if this should be a conditional value depending on the text size.
2. I adjusted the height of the stats menu to fit everything, but **I need help with the width**. The `distance travelled` especially is very squashed and overflows at 3-4 digit values.
3. **Need help** with config tool translations for non-English languages.
4. I renamed some of the stat variable names for TR1 such as `distance` to `distance travelled` for clarity. Especially since `distance` is used a lot in the codebase for distance calcs. Lmk if that's ok. i figured if we like them we can rename them in TR2 when this gets merged to libtrx and before we do TR2 BSON saves.

Default compass 1.0 text scaling:
![20250319_010834_Lost_Valley](https://github.com/user-attachments/assets/22b018b9-de26-44a2-adbd-0133c5003751)

Max compass text scaling:
![20250319_010924_Tomb_of_Qualopec](https://github.com/user-attachments/assets/bfcf7fbe-8448-4e38-a58f-558088831404)

Max text scaling end of level:
![20250319_003855_Cut_Scene_1](https://github.com/user-attachments/assets/d438a5c0-ef6b-4a9e-a38c-48e93650d9ba)

Overflow of distance travelled:
![20250319_005238_City_of_Vilcabamba](https://github.com/user-attachments/assets/7d7693bf-3a05-43f6-91f7-4d38d109fe18)

Final stats:
(PSA my final stats are slightly off because I still have Lahm's block test level as my level 1. Haven't changed it back yet lol.)
![20250319_005406_The_Great_Pyramid](https://github.com/user-attachments/assets/dd5e8782-0eb4-4991-9408-d467c57efa63)

Final stats at max text scaling:
![20250319_005444_The_Great_Pyramid](https://github.com/user-attachments/assets/5f2e03d3-cb49-4209-966a-7b9137794d0e)

walkawayy committed:
tr1/stats: add extra stats from TR2

Resolves #2651.
Added stats are ammo hits/used, health packs used, and distance
travelled.

Richard-L commented:
This is amazing! Thanks a lot for doing this. 

I'll try to make time to suggest something re overflow management (if you like). But if I can ask a dumb question to start: what does "ammo hits/used" mean? Is "used" the overall shots fired, and "ammo hits" those that actually reached a target?

rr- commented on src/tr1/game/gun/gun_misc.c:
```diff
@@ -431,6 +431,7 @@ int32_t Gun_FireWeapon(
     }
 
     ammo->ammo--;
+    Savegame_GetCurrentInfo(Game_GetCurrentLevel())->stats.ammo_used++;
```
Let's avoid changing the state on a result of a function call.
Eg
```diff
-Savegame_GetCurrentInfo(Game_GetCurrentLevel())->stats.ammo_used++;
+const …* const current_info = Savegame_GetCurrentInfo(Game_GetCurrentLevel());
+current_info->stats.ammo_used++
```
I'd also suggest we introduced `Stats_IncrementAmmoUsed()` and `Stats_IncrementAmmoHits()`.

walkawayy replied:
Ok. I was following existing convention but I'll adjust all of them.

rr- commented on src/tr1/game/lara/common.c:
```diff
@@ -333,6 +333,14 @@ void Lara_Control(void)
+            + SQUARE(item->pos.y - g_Lara.last_pos.y)
+            + SQUARE(item->pos.x - g_Lara.last_pos.x));
+
+    g_Lara.last_pos = item->pos;
```
Likewise I think it'd be nice to have `Stats_UpdateDistanceTraveled()`. I'd calculate the diff inside to reduce the strain on the callers.

rr- commented on src/tr1/game/lara/common.c:
```diff
@@ -414,6 +424,8 @@ void Lara_UseItem(const GAME_OBJECT_ID obj_id)
         CLAMPG(g_LaraItem->hit_points, LARA_MAX_HITPOINTS);
         Inv_RemoveItem(O_BIGMEDI_ITEM);
         Sound_Effect(SFX_MENU_MEDI, nullptr, SPM_ALWAYS);
+        Savegame_GetCurrentInfo(Game_GetCurrentLevel())->stats.medipacks_used +=
```
Ditto for both kinds of medipacks.

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -1445,6 +1477,13 @@ bool Savegame_BSON_LoadOnlyResumeInfo(MYFILE *fp, GAME_INFO *game_info)
+    File_Seek(fp, 0, FILE_SEEK_SET);
+    File_ReadData(fp, &header, sizeof(SAVEGAME_BSON_HEADER));
+    File_Seek(fp, 0, FILE_SEEK_SET);
+
```
We can collect the version from `M_ParseFromFile` by passing a pointer to a local variable instead of `nullptr` as the second argument:
```diff

-    // Read savegame version
-    SAVEGAME_BSON_HEADER header;
-    File_Seek(fp, 0, FILE_SEEK_SET);
-    File_ReadData(fp, &header, sizeof(SAVEGAME_BSON_HEADER));
-    File_Seek(fp, 0, FILE_SEEK_SET);
-
-    JSON_VALUE *root = M_ParseFromFile(fp, nullptr);
+    int32_t version;
+    JSON_VALUE *root = M_ParseFromFile(fp, &version);
```

rr- commented on src/tr1/game/savegame/savegame_bson.c:
```diff
@@ -286,6 +288,17 @@ static bool M_LoadResumeInfo(JSON_ARRAY *resume_arr, RESUME_INFO *resume_info)
             resume_obj, "max_kills", resume->stats.max_kill_count);
         resume->stats.max_pickup_count = JSON_ObjectGetInt(
             resume_obj, "max_pickups", resume->stats.max_pickup_count);
+        if (header_version >= VERSION_7) {
```
Let's update the comment under `VERSION_7` enum definition, to list the changes from this PR as part of its changelog.

rr- reviewed (commented):
Re: translations – other than automated tools, this isn't something I can help with. My personal workflow currently is to supply an automated translation and hope that outside contributors pick up on any weirdness. I think it's fine considering we're not officially supporting localization yet.

Re: stats dialog width – this can be controlled by changing the argument to the outermost stack's `UI_Stack_Create` call from the currently hardcoded 200 to let's say 220 (not tested).

Good call on `distance_travelled` variable name and making it unique from regular `distance`s.

My spell checker underlines `travelled` and wants me to use `traveled`. Should we act on this?

Thank you for doing this. It looks very good :relieved:

walkawayy commented:
Thanks for the width help and other comments. I'll try to make those changes tonight. I was just gonna use Google Translate, but since we had some recent PRs to fix language quirks, I wasn't sure if we wanted to keep them accurate going forward.

As far as `travelled` vs `traveled`, TR2+ uses `travelled` because the double l is a British spelling and `traveled` is the American spelling. Similar to `cancelled` vs `canceled`. I kept the British spelling since Core was British and that's how it's spelled in TR2+.

https://www.merriam-webster.com/grammar/traveling-vs-travelling-usage

lahm86 commented:
I suppose we could use a horizontal outer stack similar to the photo mode UI to get better alignment of the columns, plus text-responsiveness, but because of the intertwined centre-layout end-level stats, this might become quite unwieldy. Perhaps best tackled later if the UI is merged more with TR2, and then maybe the TR1 end-level layout could be split into its own.

Looks great otherwise 👍

rr- commented:
> I suppose we could use a horizontal outer stack similar to the photo mode UI to get better alignment of the columns, plus text-responsiveness, but because of the intertwined centre-layout end-level stats, this might become quite unwieldy. Perhaps best tackled later if the UI is merged more with TR2, and then maybe the TR1 end-level layout could be split into its own.
> 
> Looks great otherwise 👍

Yeah let's do the heavier changes once the ui apis get simpler.

Richard-L commented:
Comparison
![20250319_232329_Return_to_Egypt](https://github.com/user-attachments/assets/c1efbddb-0b97-48cb-8495-10aaabf2e77a)
![Tomb Raider II - Starring Lara Croft (USA) 2025-03-19-23-26-56](https://github.com/user-attachments/assets/d100cc25-fdc8-4c30-b7a6-054e4da7908c)

- I had PS1 style menus checked in the config tool. Maybe for this updated stats screen it's not available yet?
- Are you planning to do the left and right align instead of center?
- I wouldn't combine "Ammo Used" and "Hits" into one as it's too confusing.

walkawayy commented:
1 and 2. The TR1X version is the PS1 UI style. You posted TR2 style. Porting TR2 style end of level stats would be a different issue.

![image](https://github.com/user-attachments/assets/05fa087a-1f0d-4fac-b780-b20b097d9a28)

3. Basketball has free throws made / free throws attempted. Ex: 8-10. Same with 3 pointers. Ex: 3/6. Baseball hitting percentage has the same idea just converted to a percentage. Also the same with literally anything like a movie review is 3/5 stars. I'm not sure what is confusing. It also reduces the number of rows by 1 which is useful. The remaster also combines it.

As for the size, see the pic in `Max text scaling end of level:` in the PR to see that this is the most vertical spacing we can do to accommodate max text scaling.

walkawayy committed:
tr1/stats: introduce stat incrementing functions

walkawayy committed:
tr1/savegame: use simpler method to get savegame header version

walkawayy commented:
Force pushed some changes:
* Added translations
* Added helper functions for stat updates
* Simplified savegame header checking
* Made deaths last in the detailed stats menu. I preserved the order for the normal compass stats. Lmk if I should change it there too @rr- 
* Widened the compass menu to fit distance travelled

rr- commented on src/tr1/game/stats/common.c:
```diff
@@ -267,3 +272,59 @@ void Stats_UpdateSecrets(LEVEL_STATS *const stats)
+    current_info->stats.ammo_used++;
+}
+
+void Stats_AddMedipacksUsed(double medipack_value)
```
Nitpick: const

rr- reviewed (commented):
> Force pushed some changes:
> 
> * Added translations
> * Added helper functions for stat updates
> * Simplified savegame header checking
> * Made deaths last in the detailed stats menu. I preserved the order for the normal compass stats. Lmk if I should change it there too @rr-
> * Widened the compass menu to fit distance travelled

Thanks. Let's make the compass consistent. I'll take a look later today to check out the look'n'feel (I'm pondering if we should force the bordered variant for the extended stats), but the code LGTM :)

Richard-L commented:
@walkawayy sorry I mixed up the TR1/2 PS1 stats screen. Too used to being in TR2 of late. Could've sworn the super nice window heading and framing you'd added for the TR1 PS1 UI feature was also part of the stats screen. But looks like it's not, so a future feature issue might be "offer TR2 PS1 stats screen style in TR1".

On this occasion, for what my personal opinion/preference is worth, I think once the data exceeds the few lines that were originally showing, it's becoming increasingly more difficult to take the stats in. I would reckon this might have been why they changed it to left/right text alignment in TR2. If you're not interested in doing a full TR2 style presentation yet, perhaps would adding just a _little_ bit more space between the data type and data item per row help a little?

The combo of shots fired/shots missed is also essentially a personal preference (which is okay!).

walkawayy commented:
> @walkawayy sorry I mixed up the TR1/2 PS1 stats screen. Too used to being in TR2 of late. Could've sworn the super nice window heading and framing you'd added for the TR1 PS1 UI feature was also part of the stats screen. But looks like it's not, so a future feature issue might be "offer TR2 PS1 stats screen style in TR1".
> 
> On this occasion, for what my personal opinion/preference is worth, I think once the data exceeds the few lines that were originally showing, it's becoming increasingly more difficult to take the stats in. I would reckon this might have been why they changed it to left/right text alignment in TR2. If you're not interested in doing a full TR2 style presentation yet, perhaps would adding just a _little_ bit more space between the data type and data item per row help a little?
> 
> The combo of shots fired/shots missed is also essentially a personal preference (which is okay!).

Said this earlier but as for the size, see the pic in max text scaling end of level in the PR. This is the most vertical spacing we can do to accommodate max text scaling. I agree I think a TR2 style bordered menu could be nice. But it is technically non-OG behavior. The remasters switched to a menu like this. They also combined ammo used/hits into one line but listed used first oddly. TR2 just called the row of stats `Hits` which IMO is more confusing.

walkawayy commented:
@aredfan could you check the `distance travelled` stat in TR1 when testing for this. It seems to be behaving oddly but it's implemented the same there. Maybe it's related to turbo cheat and `/tp` command stuff though. I need to see how it behaves in TR2 so I'll try to check that.

walkawayy committed:
tr1/stats: move Deaths stat to last in non-detailed stats mode

aredfan commented:
@walkawayy the distance travelled stat in this PR seems good to me, maybe I'm not seeing the issue you're experiencing.

It's worth mentioning that the distance travelled stat increments when using the teleport cheat in TR2. That's been the case since the teleport cheat was added in [0.2](https://github.com/LostArtefacts/TR2X/releases/tag/0.2).

rr- commented:
I played with the stats a bit and I think our best option would be to merge the "detailed stats" option to also apply the TR2 style:
![20250320_183440_St_Francis_Folly](https://github.com/user-attachments/assets/eab10476-04d6-4735-9f83-27a320e27cc0)
![20250320_183444_St_Francis_Folly](https://github.com/user-attachments/assets/44163ffd-fabb-4d9a-accf-bd4e11d56c90)

```diff
diff --git a/src/libtrx/game/game_flow/sequencer_events.c b/src/libtrx/game/game_flow/sequencer_events.c
index 77744cdd4..51dc5edb5 100644
--- a/src/libtrx/game/game_flow/sequencer_events.c
+++ b/src/libtrx/game/game_flow/sequencer_events.c
@@ -99,12 +99,17 @@ static DECLARE_GF_EVENT_HANDLER(M_HandleLevelStats)
     if (seq_ctx != GFSC_NORMAL) {
         return gf_cmd;
     }
+#if TR_VERSION == 1
+    const bool use_bare_style = !g_Config.gameplay.enable_detailed_stats;
+#else
+    const bool use_bare_style = false;
+#endif
     PHASE *const phase = Phase_Stats_Create((PHASE_STATS_ARGS) {
         .background_type =
             (TR_VERSION == 1 || Game_IsInGym()) ? BK_TRANSPARENT : BK_OBJECT,
         .level_num = -1,
         .show_final_stats = false,
-        .use_bare_style = TR_VERSION == 1,
+        .use_bare_style = use_bare_style,
     });
     gf_cmd = PhaseExecutor_Run(phase);
     Phase_Stats_Destroy(phase);
```

\+ it'd need tweaking this description:

>Enables showing the maximum pickup count and kill count on each level, ammo hits, ammo used, health packs used, and distance travelled.

It could also be converted to an enum where we have 3 options – bare (OG TR1), detailed (OG TR1 + max counts), full (TR2/3 style + new rows).

LMK your thoughts. The extended stats aren't OG so I think it's OK if we tweak how they look to make them look nice, especially since its forwards compatible with the direct OG sequel.

Richard-L commented:
Yes please to the TR2 style @rr- suggested there. Looks great!

Even better if it can also work with the PS1 UI style the menus otherwise have (if chosen).

walkawayy committed:
tr1/stats: change the detailed stats option to three modes

Resolves #2658.
The three stat options include:
minimal: kills, pickups, secrets, time
detailed: minimal and max pickup and max kill count
full: detailed and ammo hits/used, health packs used, distance

walkawayy committed:
stats: fix the distance travelled display when over 1000m

Resolves #2659.

walkawayy commented:
Sorry all I think this needs one more final review. The detailed stats bool option was changed to 3 enums: minimal, detailed, and full.

I also pushed a separate commit fix for the distance travelled stat display bug when over 1000m.

lahm86 commented on docs/tr1/CHANGELOG.md:
```diff
@@ -25,12 +30,14 @@
 - fixed the `/pos` command not showing demo and cutscene titles
 - fixed the embedded bats fix causing problems inside rooms with trapdoors (regression from 4.6)
 - fixed cutscene music looping (#2591, regression from 4.8)
+- fixed the distance travelled stat displaying the wrong value when over 1000m (#2659)
```
This isn't needed for TR1 as it's unreleased, but the line is needed in TR2's readme.

aredfan reviewed (approved):
The different stats options and the bugfix both LGTM. 👍

Kappa971 commented:
@walkawayy maybe it went unnoticed but could you change the text like this? (see above)
```
"Title": "Modalità di visualizzazione delle statistiche",
"Description": "Consente di scegliere tra vari livelli di dettaglio per la visualizzazione delle statistiche.\n- Minimo: mostra le uccisioni, gli oggetti raccolti, i segreti e il tempo impiegato per completare il livello.\n- Dettagliato: mostra tutte le statistiche della modalità 'Minimo' e, in aggiunta, il numero massimo di uccisioni e di oggetti recuperabili di ogni livello.\n- Completo: mostra tutte le statistiche della modalità 'Dettagliato' e, in aggiunta, i colpi andati a segno, le munizioni utilizzate, i kit medici adoperati e la distanza percorsa."
```

walkawayy commented:
> Modalità di visualizzazione delle statistiche

Sure thanks. I only did online translation with the hopes of translation improvements in the future. I'll make this change and then merge. Hopefully Spanish and French will get improvements later.
